### ISSUE TITLE
Updating references

### ISSUE THREAD

**Issue by ibacher** [OpenConceptLab/ocl_issues#784]
**TLDR** it seems that `DELETE` to a collections references endpoint is not honouring the `cascade=sourcemappings` parameter.

----

So, I'm not sure if this is an issue with how the OCL for OpenMRS client uses the API or the API itself, but we seem to have a bug which is making it so that updated mappings show up in the UI but do not show up in the export because of an issue updating the reference.

When editing a custom concept in the OCL for OpenMRS client, we send the following requests:

```
PUT https://api.openconceptlab.org/<owner path>/sources/<source id>/concepts/<concept id>
```

For each mapping:

```
PUT https://api.openconceptlab.org/<owner path>/sources/<source id>/mappings/<mapping id>
```

Finally, to update the references we do:

```
DELETE https://api.openconceptlab.org/<owner path>/collections/<collection id>/references/?cascade=sourcemappings
```

Followed by:

```
PUT https://api.openconceptlab.org/<owner path>/collections/<collection id>/references/?cascade=sourcemappings
```

All of these appear to succeed, i.e., return a HTTP 200 response, except the `DELETE` which returns 204.

However, digging a little deeper, I see that the last `PUT` request actually fails to update the references appropriately; more exactly, it successfully updates the *concept* but not any of the concept mappings. Despite the HTTP 200 response, I see errors when trying to add the mappings reference that say "Concept or Mapping reference name must be unique in a collection."

For a concrete example of this, just try editing and saving [the concept here](https://openmrs.openconceptlab.org/users/ibacher/sources/IBT/concepts/4000/259120/?linkedDictionary=/users/ibacher/collections/IBT/).

When I play around a bit with this in Postman, I notice that the `DELETE` request seems to be superfluous, i.e., nothing gets deleted from the references for my collection, which seems to be an error. What I don't understand is why the reference for the concept is successfully updated in the final `PUT` request but not the mappings associated with that concept.

This is a major issue for us because it means that the export of the collection will never have any modifications to concept mappings after the initial mapping has been created.

**Comment by snyaggarwal** [comment]
Hi @ibacher,
I tried to reproduce this but couldn't. Things are working as expected. 
This is what I tried:

1. PUT http://127.0.0.1:8000/users/ocladmin/collections/c5/references/?cascade=sourcemappings
    {"data": {"expressions": ["/users/ocladmin/sources/test2/concepts/c1/"]}}
2. Above one created two references:
  - Concept Reference `/users/ocladmin/sources/test2/concepts/c1/1273` (latest version)
  - Mapping Reference `/users/ocladmin/sources/test2/mappings/m1/12545/` (related mapping to above concept)
3. DELETE http://127.0.0.1:8000/users/ocladmin/collections/c5/references/?cascade=sourcemappings
    {"expressions": ["/users/ocladmin/sources/test2/concepts/c1/1273/"]}
4. Above deleted both the references.

@ibacher Can you provide more info on what payload you are trying with PUT/DELETE calls?

**Comment by ibacher** [comment]
Hi @snyaggarwal!

Thanks for trying that out. I should've put more details in. So I have a test collection in production (IBT). The references for this look this:

```
GET https://api.openconceptlab.org/users/ibacher/collections/IBT/references/
```

```json
[
    {
        "expression": "/users/ibacher/sources/IBT/concepts/4000/259122/",
        "reference_type": "concepts",
        "id": 165288
    },
    {
        "expression": "/users/ibacher/sources/IBT/mappings/462408/462420/",
        "reference_type": "mappings",
        "id": 165285
    },
    {
        "expression": "/users/ibacher/sources/IBT/mappings/462416/462417/",
        "reference_type": "mappings",
        "id": 165239
    }
]
```

This represents a single concept with two mappings, however, these two mappings are slightly out of date, because the mappings in my source have been updated (which happened in step 2 of the process I outlined in the beginning), so the latest version of mapping 462408 in my source is version 462423, and for mapping 462416 the version is 462422.

Then I send the delete request with this payload:

```
DELETE https://api.openconceptlab.org/users/ibacher/collections/IBT/references/?cascade=sourcemappings
```

```json
{
    "references": ["/users/ibacher/sources/IBT/concepts/4000/259122/"]
}
```

Which gets a 204 response, but when I run the GET request after that delete, I see that the concept has been removed, but the mappings have not. (I also tried changing `references` to `expressions` as per your example above, but that doesn't change anything).

If I change the body of the delete request to:

```json
{
    "references": ["/users/ibacher/sources/IBT/concepts/4000/259122/", "/users/ibacher/sources/IBT/mappings/462408/462420/", "/users/ibacher/sources/IBT/mappings/462416/462417/"]
}
```

This works, but is obviously less than ideal.

I've attached a [HAR file](https://en.wikipedia.org/wiki/HAR_(file_format)) which includes the complete playback of requests caused by editing the concept in the OCL for OpenMRS app (the only change I've made to it is to remove my OCL API token.

[openmrs.openconceptlab.org_Archive [21-06-03 10-19-54].har.log](https://github.com/OpenConceptLab/ocl_issues/files/6592309/openmrs.openconceptlab.org_Archive.21-06-03.10-19-54.har.log)

**Comment by snyaggarwal** [comment]
@ibacher Found the issue. Will be fixing this soon.

**Comment by snyaggarwal** [comment]
@ibacher This is fixed and deployed on QA. You can test it out on QA, will be proceeding with Staging/Production deployment tonight.

**Comment by snyaggarwal** [comment]
Hey @ibacher can we close this?

**Comment by ibacher** [comment]
@snyaggarwal Happy to!
